### PR TITLE
Custom sgemm

### PR DESCRIPTION
**Pull request by yekatkov**
Custom sgemm was added for process matrix multiplication on AMD CPU.

### CODE REVIEW

File: NeoMathEngine/src/CPU/x86/avx/src/MatrixMultiplying/Kernels/Kernels_6x.h [hunk 1]
```diff
@@ -0,0 +1,674 @@
+/* Copyright © 2017-2021 ABBYY Production LLC
```
Comment (SAngeliuk) [inline NeoMathEngine/src/CPU/x86/avx/src/MatrixMultiplying/Kernels/Kernels_6x.h]: One kernel - one file!

File: NeoMathEngine/src/CPU/x86/avx/src/MatrixMultiplying/Interleavers/Interleavers.h [hunk 2]
```diff
@@ -0,0 +1,444 @@
+/* Copyright © 2017-2021 ABBYY Production LLC
```
Comment (SAngeliuk) [inline NeoMathEngine/src/CPU/x86/avx/src/MatrixMultiplying/Interleavers/Interleavers.h]: One interleaver - one file?

File: NeoMathEngine/src/CPU/x86/AvxDll.cpp [hunk 3]
```diff
@@ -138,6 +138,7 @@ bool CAvxDll::loadFunctions()
 
 bool CAvxDll::isAvxAvailable()
 {
+	// TODO: move to CPUInfo
```
Comment (SAngeliuk) [inline NeoMathEngine/src/CPU/x86/AvxDll.cpp]: Let's do it!

File: NeoMathEngine/include/NeoMathEngine/SimdMathEngine.h [hunk 4]
```diff
@@ -32,6 +39,8 @@ class ISimdMathEngine : public CCrtAllocatedObject {
 
 	virtual void BlobConvolution( const CConvolutionDesc& convDesc, const float* source,
 		const float* filter, const float* freeTerm, float* result ) const = 0;
+
+	virtual SgemmFunc GetSgemmFunction() const = 0;
```
Comment (SAngeliuk) [inline NeoMathEngine/include/NeoMathEngine/SimdMathEngine.h:43]: Why Get...? Why not virtual void Sgemm(...) const = 0 ?

File: NeoMathEngine/src/CPU/x86/avx/src/MatrixMultiplying/Interleavers/Interleavers.h (hunk 2 above)
Comment (yekatkov) [reply NeoMathEngine/src/CPU/x86/avx/src/MatrixMultiplying/Interleavers/Interleavers.h]: All ARM's Interleavers are placed in single file.

File: NeoMathEngine/src/CPU/x86/avx/src/MatrixMultiplying/Kernels/Kernels_6x.h (hunk 1 above)
Comment (yekatkov) [reply NeoMathEngine/src/CPU/x86/avx/src/MatrixMultiplying/Kernels/Kernels_6x.h]: Done

File: NeoMathEngine/src/CPU/x86/AvxDll.cpp (hunk 3 above)
Comment (yekatkov) [reply NeoMathEngine/src/CPU/x86/AvxDll.cpp]: Done

File: NeoMathEngine/include/NeoMathEngine/SimdMathEngine.h (hunk 4 above)
Comment (yekatkov) [reply NeoMathEngine/include/NeoMathEngine/SimdMathEngine.h:43]: _**virtual void Sgemm**_ doesn't bring clarity in calling code.

File: NeoMathEngine/src/CPU/CpuMathEngine.h [hunk 5]
```diff
@@ -489,6 +490,7 @@ class CCpuMathEngine : public IMathEngine, public IRawMemoryManager {
 
 	CDllLoader dllLoader; // loading library for simd instructions
 	std::unique_ptr<const ISimdMathEngine> simdMathEngine; // interface for using simd instructions
+	SgemmFunc CustomSgemmFunction; // Used when it is availabled and is faster then default sgemm
```
Comment (SAngeliuk) [inline NeoMathEngine/src/CPU/CpuMathEngine.h]: CustomSgemmFunction --> **c**ustomSgemmFunction

File: NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp [hunk 6]
```diff
@@ -491,6 +491,8 @@ void CCpuMathEngine::blobConvolutionForwardAlgo1( const CCpuConvolutionDesc& des
 					multiplyMatrixByTransposedMatrixAndAdd( tempBlobPtr, result.Height() * resultCount, filter.ObjectSize(),
 						filter.ObjectSize(), filterData, filter.BatchWidth(), filter.ObjectSize(), outputTransposedPtr,
 						filter.BatchWidth() );
+
```
Comment (SAngeliuk) [inline NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp]: Why?)

File: NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp [hunk 7]
```diff
@@ -1037,7 +1039,7 @@ void CCpuMathEngine::blobConvolutionLearnAlgo2( const CCpuConvolutionDesc& desc,
 					inputMatrix,
 					input.Depth() * input.Channels(),
 					input.Depth() * input.Channels(),
-					filterMatrix, filterDiff.ObjectSize() );
+					filterMatrix, filterDiff.ObjectSize());
```
Comment (SAngeliuk) [inline NeoMathEngine/src/CPU/CpuMathEngineDnnConv.cpp]: Why?)

File: NeoMathEngine/src/CPU/arm/CpuArmMathEngineMatrixMultiplying.cpp [hunk 8]
```diff
@@ -24,6 +24,7 @@ limitations under the License.
 #include <MatrixMultiplyingInterleavedCommon/MatrixMultiplying.h>
 #include <MatrixMultiplyingInterleavedCommon/CpuMemoryHelper.h>
 #include <MathEngineCommon.h>
+#include <CpuMathEngine.h>
```
Comment (SAngeliuk) [inline NeoMathEngine/src/CPU/arm/CpuArmMathEngineMatrixMultiplying.cpp:27]: Why is it needed?

File: NeoMathEngine/src/CPU/x86/CpuX86MathEngineBlasMkl.cpp [hunk 9]
```diff
@@ -69,43 +76,59 @@ void CCpuMathEngine::multiplyMatrixByMatrixAndAdd( const float* first, int first
...
+		MultiplyMatrix<false, false, CTmpMemoryHandler>( this, CpuInfo, first, firstRowSize, second, secondRowSize,
+			result, resultRowSize, firstHeight, secondWidth, firstWidth );
 #endif
+	}
 }
 
 void CCpuMathEngine::multiplyMatrixByTransposedMatrix(const float* first, int firstHeight,
 	int firstWidth, int firstRowSize, const float* second, int secondHeight, int secondRowSize,
-	float* result, int resultRowSize)
+	float* result, int resultRowSize )
```
Comment (SAngeliuk) [inline NeoMathEngine/src/CPU/x86/CpuX86MathEngineBlasMkl.cpp]: why?

File: NeoMathEngine/src/CPU/x86/CpuX86MathEngineBlasMkl.cpp [hunk 10]
```diff
@@ -201,36 +224,50 @@ void CCpuMathEngine::MultiplyTransposedMatrixBySparseMatrixAndAdd( int firstHeig
 
 void CCpuMathEngine::multiplyTransposedMatrixByMatrix(const float* first, int firstHeight,
 	int firstWidth, const float* second, int secondWidth,
-	float* result)
+	float* result )
```
Comment (SAngeliuk) [inline NeoMathEngine/src/CPU/x86/CpuX86MathEngineBlasMkl.cpp]: why?

File: NeoMathEngine/src/CPU/x86/CpuX86MathEngineBlasMkl.cpp [hunk 11]
```diff
@@ -201,36 +224,50 @@ void CCpuMathEngine::MultiplyTransposedMatrixBySparseMatrixAndAdd( int firstHeig
...
+			result, resultRowSize, firstWidth, secondWidth, firstHeight );
 #endif
+	}
 }
 
 void CCpuMathEngine::multiplyTransposedMatrixByMatrixAndAdd(const float* first,
 	int firstHeight, int firstWidth, int firstRowSize,
 	const float* second, int secondWidth, int secondRowSize,
-	float* result, int resultRowSize)
+	float* result, int resultRowSize )
```
Comment (SAngeliuk) [inline NeoMathEngine/src/CPU/x86/CpuX86MathEngineBlasMkl.cpp]: ?

**Review by SAngeliuk — APPROVED** [review]

**Pull request merged**